### ISSUE TITLE
OpenCV DNN object detection returns always the same calssid

### ISSUE THREAD

**Issue by garutilorenzo** [ultralytics/ultralytics#2844]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everybody, and thank you for this awesome too! 😄 

I've successfully trained my custom dataset

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8s.pt')  # load a pretrained model (recommended for training)

# Train the model
model.train(
    name='MyModel',
    data='/root/work/images/data.yaml', 
    epochs=100, 
    imgsz=640
)
```

and if I try to find some objects inside an image I get the right results:

```
python3 yolo_test.py

cat
[ 344  808  572 1208]
[ 458 1008  227  400]
[    0.84057]
```

this is the content of yolo_test.py:

```python
from ultralytics import YOLO
from PIL import Image
import requests
from io import BytesIO
import cv2
import numpy as np

model = YOLO('/root/work/MyModel/weights/best.pt')

inputs = ['https://example.com/cat.jpg']

results = model(inputs)
for result in results:
    boxes = result.boxes  # Boxes object for bbox outputs
    masks = result.masks  # Masks object for segmentation masks outputs
    probs = result.probs  # Class probabilities for classification outputs
    
    # print(boxes.xyxy)  # box with xyxy format, (N, 4)
    # print(boxes.xywh)  # box with xywh format, (N, 4)
    # print(boxes.xyxyn)  # box with xyxy format but normalized, (N, 4)
    # print(boxes.xywhn)  # box with xywh format but normalized, (N, 4)
    # print(boxes.conf)  # confidence score, (N, 1)
    # print(boxes.cls)  # cls, (N, 1)
    # print(boxes.data)  # raw bboxes tensor, (N, 6) or boxes.boxes
    
    for box in boxes:
        r = box.xyxy[0].numpy().astype(int)
        r_normal = box.xywh[0].numpy().astype(int)
        print(result.names[int(box.cls[0])])
        print(r)
        print(r_normal)
        confidence = box.conf.numpy()
        print(confidence)
        if confidence > .7:
            area_to_crop = (r[0], r[1], r[2], r[3])
            cropped_img  = img.crop(area_to_crop)
            cropped_img.save('/root/work/cropped_img.jpg')
```

Now I've exported the model on onnx format:

```python
from ultralytics import YOLO
from PIL import Image
import requests
from io import BytesIO
import cv2
import numpy as np

model = YOLO('/root/work/MyModel/weights/best.pt')
# Export the model
model.export(format='onnx', opset=12)
```

and I've tried to use opencv for the object detection this is my 'predict_cv2.py':

```python
import cv2
import numpy as np
import requests
from PIL import Image

# Load Model
net = cv2.dnn.readNet('/root/work/MyModel/weights/best.onnx')

# net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
# net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA_FP16)

INPUT_WIDTH = 640
INPUT_HEIGHT = 640
# SCORE_THRESHOLD = 0.2
# NMS_THRESHOLD = 0.4
# CONFIDENCE_THRESHOLD = 0.4

CLASESS_YOLO = ['dog','cat']


inputs = ['https://example.com/cat.jpg']

response = requests.get(inputs[0])
cv2_img = cv2.imdecode(np.array(bytearray(response.content), dtype=np.uint8), -1)
blob = cv2.dnn.blobFromImage(cv2_img, 1/255.0, (INPUT_WIDTH, INPUT_HEIGHT), swapRB=True, crop=False)
net.setInput(blob)
preds = net.forward()

preds = preds.transpose((0, 2, 1))

# Extract output detection
class_ids, confs, boxes = list(), list(), list()

image_height, image_width, _ = cv2_img.shape
x_factor = image_width / INPUT_WIDTH
y_factor = image_height / INPUT_HEIGHT

min_conf = .3

rows = preds[0].shape[0]

for i in range(rows):
    row = preds[0][i]
    conf = row[4]
    
    classes_score = row[4:]
    _,_,_, max_idx = cv2.minMaxLoc(classes_score)
    class_id = max_idx[1] 

    if (classes_score[class_id] > min_conf):
        confs.append(conf)
        label = CLASESS_YOLO[int(class_id)]
        print(class_id)
        class_ids.append(label)
        #extract boxes
        x, y, w, h = row[0].item(), row[1].item(), row[2].item(), row[3].item() 
        left = int((x - 0.5 * w) * x_factor)
        top = int((y - 0.5 * h) * y_factor)
        width = int(w * x_factor)
        height = int(h * y_factor)
        box = np.array([left, top, width, height])
        boxes.append(box)
        
r_class_ids, r_confs, r_boxes = list(), list(), list()

indexes = cv2.dnn.NMSBoxes(boxes, confs, min_conf, min_conf-0.1) 

for i in indexes:
    r_class_ids.append(class_ids[i])
    r_confs.append(confs[i])
    r_boxes.append(boxes[i])

print(r_class_ids)
print(r_confs)
print(r_boxes)

for i in indexes:
    box = boxes[i]
    left = box[0]
    top = box[1]
    width = box[2]
    height = box[3]
    
    cv2.rectangle(cv2_img, (left, top), (left + width, top + height), (0,255,0), 3)
result_image = Image.fromarray(cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB))
result_image.save('/root/work/image_cv2.jpg')
```

All r_boxes, r_confs  and even the saved image '/root/work/image_cv2.jpg' are right.
The object is detected correctly. But the detected 'class_id' -> 'class_id = max_idx[1] ' is always the same '0'

This is the result of 'predict_cv2.py':

```
['dog']
[0.8799546]
[array([364, 802, 206, 416])]
```

Can someone help me?
What am I doing wrong?

Thank you!
Lorenzo

**Comment by glenn-jocher** [comment]
@garutilorenzo hello Lorenzo! We're glad to hear that you've successfully trained a custom dataset using YOLOv8. Thank you for your kind words and for sharing your experience.

Regarding your question, it seems that the issue stems from the line `class_id = max_idx[1]` in your 'predict_cv2.py' file. Instead of using `max_idx[1]` to get the index of the maximum value, you should use `np.argmax(classes_score)` to get the index of the maximum value in 'classes_score'. The `max_idx[1]` line only gives you the y-coordinate of the maximum value in 'classes_score', which does not correspond to the correct class ID.

I hope this helps you solve the issue you were experiencing. If you have any further questions or if this doesn't solve your problem, please don't hesitate to ask.

**Comment by garutilorenzo** [comment]
Hi @glenn-jocher  thank you for the quick response. Unfortunately `np.argmax(classes_score)`  does not solve the problem:

```
this is the content of classes_score:
[0.52030647 0.04857335]
this the full content of max_id:
(0, 0)
this is the output of np.argmax(classes_score):
0
0
```

**Comment by garutilorenzo** [comment]
Hi @glenn-jocher I think I've found the solution. I'm currentrly train a new model but from the first test it seems that I need to map a `"dummy class"` on the data.yaml file.

This was the original data.yaml:

```yaml
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 2
names: ['dog', 'cat']
```

and this is the new one:

```yaml
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 3
names: ['-', 'dog', 'cat']
```

Then I've fixed all the labels .txt files with the new class mapping.
I will share the new results when the new model is available

**Comment by glenn-jocher** [comment]
Hi @garutilorenzo, glad to hear that you may have found a solution to your problem. It seems that you needed to add a `"dummy class"` to your data.yaml file with the code '-'. The reason for this is likely because the class index expected by YOLOv8 starts at 0. Therefore, all of your class indices were likely off by 1 since 'dog' and 'cat' have indices 1 and 2, respectively. 

By adding a dummy class with an index of 0, you shifted all of your other classes up by one index, which should make it easier to map the correct class indices in your code. It's important to note that you'll need to update all of your label files to use the new class index values. 

I hope this solves your problem and that your new model performs well! Don't hesitate to reach out if you have any more questions.

**Comment by garutilorenzo** [comment]
Ok so I can confirm that the problem was the class index on the data.yaml.  Adding the 0 index solves the problem.
For future generations, I've made some changes on `predict_cv2.py`:

```python
import cv2
import numpy as np
import requests
from PIL import Image

# Load Model
net = cv2.dnn.readNet('/root/work/MyModel/weights/best.onnx')

# net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
# net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA_FP16)

INPUT_WIDTH = 640
INPUT_HEIGHT = 640
# SCORE_THRESHOLD = 0.2
# NMS_THRESHOLD = 0.4
# CONFIDENCE_THRESHOLD = 0.4

CLASESS_YOLO = ['-','dog','cat']


inputs = ['https://example.com/cat.jpg']

response = requests.get(inputs[0])
cv2_img = cv2.imdecode(np.array(bytearray(response.content), dtype=np.uint8), -1)
blob = cv2.dnn.blobFromImage(cv2_img, 1/255.0, (INPUT_WIDTH, INPUT_HEIGHT), swapRB=True, crop=False)
net.setInput(blob)
preds = net.forward()

preds = preds.transpose((0, 2, 1))

# Extract output detection
class_ids, confs, boxes = list(), list(), list()

image_height, image_width, _ = cv2_img.shape
x_factor = image_width / INPUT_WIDTH
y_factor = image_height / INPUT_HEIGHT

min_conf = .3

rows = preds[0].shape[0]

for i in range(rows):
    row = preds[0][i]
    conf = row[4]
    
    classes_score = row[4:]
    minVal,maxVal,min_idx,max_idx = cv2.minMaxLoc(classes_score)
    class_id = max_idx[1] 

    if (classes_score[class_id] > min_conf):
        print('maxVal:')
        print(maxVal)
        print('conf:')
        print(conf)
        # confs.append(conf)
        confs.append(maxVal) # <- use maxVal instead of conf. conf it's a very small number
        label = CLASESS_YOLO[int(class_id)]
        class_ids.append(label)
        #extract boxes
        x, y, w, h = row[0].item(), row[1].item(), row[2].item(), row[3].item() 
        left = int((x - 0.5 * w) * x_factor)
        top = int((y - 0.5 * h) * y_factor)
        width = int(w * x_factor)
        height = int(h * y_factor)
        box = np.array([left, top, width, height])
        boxes.append(box)
        
r_class_ids, r_confs, r_boxes = list(), list(), list()

indexes = cv2.dnn.NMSBoxes(boxes, confs, min_conf, min_conf-0.1) 

for i in indexes:
    r_class_ids.append(class_ids[i])
    r_confs.append(confs[i])
    r_boxes.append(boxes[i])

print(r_class_ids)
print(r_confs)
print(r_boxes)

for i in indexes:
    box = boxes[i]
    left = box[0]
    top = box[1]
    width = box[2]
    height = box[3]
    
    cv2.rectangle(cv2_img, (left, top), (left + width, top + height), (0,255,0), 3)
result_image = Image.fromarray(cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB))
result_image.save('/root/work/image_cv2.jpg')
```

I use the `maxVal` instead of  `conf` to populate the `confs` list. `conf` contains a very small number. This is an example output:

```
time python3 predict_cv2.py 
maxVal:
0.7962904572486877
conf:
3.4137877e-06
maxVal:
0.7295417785644531
conf:
3.8158214e-06
maxVal:
0.6139293313026428
conf:
3.6606605e-06
maxVal:
0.6775078773498535
conf:
3.1381107e-06
maxVal:
0.683744490146637
conf:
3.0724323e-06
maxVal:
0.5553311705589294
conf:
5.5505593e-06
maxVal:
0.638124942779541
conf:
4.102716e-06
maxVal:
0.6537102460861206
conf:
3.1637428e-06
maxVal:
0.7355085015296936
conf:
4.1622325e-06
maxVal:
0.662242591381073
conf:
3.875756e-06
['cat']
[0.7962904572486877]
[array([257, 105, 166, 227])]

real    0m1.926s
user    0m2.884s
sys     0m0.802s
```

I've alsto created a `predict_onx.py` that uses `onnxruntime` module for detecting the objects:

```python
import os
from io import BytesIO
import requests
import cv2
import numpy as np
from PIL import Image
import onnxruntime

opt_session = onnxruntime.SessionOptions()
opt_session.enable_mem_pattern = False
opt_session.enable_cpu_mem_arena = False
opt_session.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL

model_path = '/root/work/MyModel/weights/best.onnx'
ort_session = onnxruntime.InferenceSession(model_path)

model_inputs = ort_session.get_inputs()
input_names = [model_inputs[i].name for i in range(len(model_inputs))]
input_shape = model_inputs[0].shape
print(input_shape)

model_output = ort_session.get_outputs()
output_names = [model_output[i].name for i in range(len(model_output))]


inputs = ['https://example.com/cat.jpg']

response = requests.get(inputs[0])
cv2_img = cv2.imdecode(np.array(bytearray(response.content), dtype=np.uint8), -1)

image_height, image_width = cv2_img.shape[:2]
Image.fromarray(cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB))
input_height, input_width = input_shape[2:]

image_rgb = cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB)
resized = cv2.resize(image_rgb, (input_width, input_height))

# Scale input pixel value to 0 to 1
input_image = resized / 255.0
input_image = input_image.transpose(2,0,1)
input_tensor = input_image[np.newaxis, :, :, :].astype(np.float32)
print(input_tensor.shape)

outputs = ort_session.run(output_names, {input_names[0]: input_tensor})[0]

predictions = np.squeeze(outputs).T
conf_thresold = .6
# Filter out object confidence scores below threshold
scores = np.max(predictions[:, 4:], axis=1)
predictions = predictions[scores > conf_thresold, :]
scores = scores[scores > conf_thresold]

class_ids = np.argmax(predictions[:, 4:], axis=1)
boxes = predictions[:, :4]

#rescale box
input_shape = np.array([input_width, input_height, input_width, input_height])
boxes = np.divide(boxes, input_shape, dtype=np.float32)
boxes *= np.array([image_width, image_height, image_width, image_height])
boxes = boxes.astype(np.int32)

# Apply non-maxima suppression to suppress weak, overlapping bounding boxes
indices = cv2.dnn.NMSBoxes(boxes, scores, conf_thresold, conf_thresold-0.1)

CLASSES = ['-','dog','cat']

def xywh2xyxy(x):
    # Convert bounding box (x, y, w, h) to bounding box (x1, y1, x2, y2)
    y = np.copy(x)
    y[..., 0] = x[..., 0] - x[..., 2] / 2
    y[..., 1] = x[..., 1] - x[..., 3] / 2
    y[..., 2] = x[..., 0] + x[..., 2] / 2
    y[..., 3] = x[..., 1] + x[..., 3] / 2
    return y

image_draw = cv2_img.copy()
for (bbox, score, label) in zip(xywh2xyxy(boxes[indices]), scores[indices], class_ids[indices]):
    bbox = bbox.round().astype(np.int32).tolist()
    cls_id = int(label)
    cls = CLASSES[cls_id]
    color = (0,255,0)
    cv2.rectangle(image_draw, tuple(bbox[:2]), tuple(bbox[2:]), color, 2)
    cv2.putText(image_draw,
                f'{cls}:{int(score*100)}', (bbox[0], bbox[1] - 2),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.60, [225, 255, 255],
                thickness=1)
result_image = Image.fromarray(cv2.cvtColor(image_draw, cv2.COLOR_BGR2RGB))
result_image.save('/root/work/image_onx2.jpg')
```

Thank you very much for the support @glenn-jocher ✌️ 😄

**Comment by glenn-jocher** [comment]
@garutilorenzo hi there! Thank you for the update on the issue. It's great to hear that adding the 0 index to your data.yaml file solved the problem in your YOLOv8 implementation.

Thank you for sharing your solution with us as well as the changes you made to the `predict_cv2.py` script. It is also useful to know that you used `maxVal` instead of `conf` to populate the `confs` list as `conf` contains a very small number.

Additionally, it's great to see that you were able to implement `predict_onx.py` with `onnxruntime` to detect objects. You use `cv2.resize` to resize the input image, then apply non-maxima suppression to suppress weak, overlapping bounding boxes. You then convert the bounding box (x, y, w, h) to bounding box (x1, y1, x2, y2) and use it to draw rectangles and text on the input image.

Thank you for your contribution and detailed response, and we're glad we could help! Let us know if there's anything else we can help with.
